test/ATokenCheckpointed.t.sol

```diff
@@ -0,0 +1,1974 @@
+
+    // Submit votes on behalf of the pool.
+    aToken.castVote(_proposalId);
+
+    (uint256 _againstVotes, uint256 _forVotes, /*uint256 _abstainVotes */ ) =
+      governor.proposalVotes(_proposalId);
+
+    // userA's vote *should* have beaten userB's.
+    assertGt(_forVotes, _againstVotes, "rebasing isn't reflected in vote weight");
+  }
+
+  function _testCannotExpressVoteAfterVotesHaveBeenCast(
+    address _userA,
+    address _userB,
+    uint8 _supportType
+  ) private {
+    // Deposit some funds.
+    _mintGovAndSupplyToAave(_userA, 1 ether);
+    _mintGovAndSupplyToAave(_userB, 1 ether);
+
+    // Advance one block so that our votes will be checkpointed by the govToken;
+    vm.roll(block.number + 1);
+
+    // Create the proposal.
+    uint256 _proposalId = _createAndSubmitProposal();
+
+    // Express voting preference on the proposal.
+    vm.prank(_userA);
+    aToken.expressVote(_proposalId, _supportType);
+
+    // Wait until after the voting period
+    vm.roll(aToken.internalVotingPeriodEnd(_proposalId) + 1);
+
+    // submit votes on behalf of the pool
+    aToken.castVote(_proposalId);
+
+    // _userB should not be able to express his/her vote on the proposal since the
+    // vote was cast.
+    vm.expectRevert(bytes("too late to express, votes already cast"));
+    vm.prank(_userB);
+    aToken.expressVote(_proposalId, _supportType);
+  }
+
+  function _testCannotCastVoteWithoutVotesExpressed(address _who, uint8 _supportType) private {
+    // Deposit some funds.
+    _mintGovAndSupplyToAave(_who, 1 ether);
+
+    // Advance one block so that our votes will be checkpointed by the govToken;
+    vm.roll(block.number + 1);
+
+    // Create the proposal.
+    uint256 _proposalId = _createAndSubmitProposal();
+
+    // Wait until after the voting period
+    vm.roll(aToken.internalVotingPeriodEnd(_proposalId) + 1);
+
+    // Try to submit votes on behalf of the pool. It should fail.
+    vm.expectRevert(bytes("no votes expressed"));
+    aToken.castVote(_proposalId);
+
+    // Express voting preference on the proposal.
+    vm.prank(_who);
+    aToken.expressVote(_proposalId, _supportType);
+
+    // Now votes should be castable.
+    aToken.castVote(_proposalId);
+  }
+
+  function _testCastVoteWithTransferredATokens(
+    address _userA,
+    address _userB,
+    uint256 _weight,
+    uint256 _transferAmount,
+    uint8 _supportTypeA,
+    uint8 _supportTypeB
+  ) private {
+    // Deposit some funds.
+    _mintGovAndSupplyToAave(_userA, _weight);
+    assertEq(aToken.balanceOf(_userA), _weight);
+    assertEq(aToken.balanceOf(_userB), 0);
+
+    // Transfer all aTokens from userA to userB.
+    vm.prank(_userA);
+    aToken.transfer(_userB, _transferAmount);
+    assertEq(aToken.balanceOf(_userA), _weight - _transferAmount);
+    assertEq(aToken.balanceOf(_userB), _transferAmount);
+
+    // Advance one block so that our votes will be checkpointed by the govToken;
+    vm.roll(block.number + 1);
```

Comment:
Oh, good point. Maybe it's not needed in this test

I had added it because you can't lookup a checkpointed value until after the block has been "mined". But proposing doesn't require looking up checkpointed values. It sets the proposal block, which can later be used to lookup the relevant voting weights.

I'll see where else I can remove this 👍

Reply:
Removed in f800bc6

Reply:
There are four other `vm.roll(block.number + 1);` usages, should those also be removed? Tried to test myself, but due to the compiler error I could not test

Reply:
One was unnecessary, removed it in 53732a8

The others are needed b/c they test our code's checkpointing of raw balances. And the way the OZ checkpointing library works, you can't lookup the checkpointed value at a block until that block is past.